### PR TITLE
Add `track_content` option to allow hashing of `include_dependency`s

### LINKED ISSUE #52462: `include()`'ing nonexistant file now gives misleading "can only hash files" error
If a package has an incorrect `include()`, it used to throw an error like this:

```
$ cat src/Example.jl | head
module Example
export hello, domath

include("this_is_nonexistant.jl")

"""
    hello(who::String)

Return "Hello, `who`".
"""
$ julia +1.9 --project -e 'using Example' 2>&1 | head -1
ERROR: LoadError: SystemError: opening file "/Users/sabae/.julia/dev/Example/src/this_is_nonexistant.jl": No such file or directory
```

Whereas:
```
$ julia +master --project -e 'using Example' 2>&1 | head -1
ERROR: LoadError: AssertionError: can only hash files
```

This appears to have changed as of https://github.com/JuliaLang/julia/commit/f2df1b41b85843120a9715449eb5ff555a4a172d

### PR DESCRIPTION
**Pull request by fatteneder**
Continuation of #49866.
Fixes #52462 

So far any `include_dependency` was tracked by `mtime`.
A package using `include_dependency` can't be relocated without recompilation if the dependency also needs to be relocated.

With `include_dependency(path, track_content=true)` the tracking works like for `include`,
i.e. recompilation is only triggered when the file's content changes.
In case `path` is a directory we use the string `join(readdir(path))` as content.

### CODE REVIEW

**Comment by fatteneder** [comment]
Need to add a Julia compat note to docs.

File: test/loading.jl [hunk 1]
```diff
@@ -1377,3 +1377,31 @@ end
...
+        open(joinpath(foo_path, "Project.toml"); write=true) do io
+            println(io, """
+            name = "Foo"
+            uuid = "00000000-0000-0000-0000-000000000001"
+            version = "1.0.0"
+            """)
+        end
+
+        file = joinpath(depot, "dev", "non-existent.jl")
+        @test_throws SystemError("opening file \"$file\"") include(file)
```
Comment (fatteneder) [inline test/loading.jl:1405]:
```suggestion
        @test_throws SystemError("opening file $(repr(file))") include(file)
```

File: base/loading.jl [hunk 2]
```diff
@@ -1738,31 +1738,34 @@ function _include_dependency(mod::Module, _path::AbstractString; track_content=t
...
+                push!(_require_dependencies, (mod, path, UInt64(0), UInt32(0), mtime(path)))
             end
         end
     end
     return path, prev
 end
 
 """
-    include_dependency(path::AbstractString)
+    include_dependency(path::AbstractString; track_content::Bool=false)
```
Comment (stevengj) [inline base/loading.jl:1753]: Shouldn't `track_content=true` be the default behavior?  Wasn't that the old behavior?

Comment (stevengj) [reply base/loading.jl:1753]: Hmm, no, I guess prior to #49866 we only looked at `mtime`?

**Review by staticfloat — APPROVED** [review]

**Comment by fatteneder** [comment]
After rebasing there is now a test failure that seems related to #53103.
Need to look into that.

File: test/loading.jl [hunk 3]
```diff
@@ -1504,3 +1504,31 @@ end
...
+    mktempdir() do depot
+        project_path = joinpath(depot, "project")
+        mkpath(project_path)
+
+        # Create a `Foo.jl` package
+        foo_path = joinpath(depot, "dev", "Foo")
+        mkpath(joinpath(foo_path, "src"))
+        open(joinpath(foo_path, "src", "Foo.jl"); write=true) do io
+            println(io, """
+            module Foo
```
Comment (vchuravy) [inline test/loading.jl:1518]:
```suggestion
        foo_path = joinpath(depot, "dev", "Foo52462")
        mkpath(joinpath(foo_path, "src"))
        open(joinpath(foo_path, "src", "Foo52462.jl"); write=true) do io
            println(io, """
            module Foo52462
```

File: test/loading.jl [hunk 4]
```diff
@@ -1504,3 +1504,31 @@ end
...
+            include("non-existent.jl")
+            end
+            """)
+        end
+        open(joinpath(foo_path, "Project.toml"); write=true) do io
+            println(io, """
+            name = "Foo"
```
Comment (vchuravy) [inline test/loading.jl:1525]:
```suggestion
            name = "Foo52462"
```

**Pull request merged**